Gradient generation

breuleux opened the pull request:
This adds gradients for various primitives, and gradient generation. The tests pass and there is 100% coverage, but it's worth mentioning the whole thing is slightly rushed. A few comments:

* There are two xfails which should be fixed at some point.
* Support for array operations has not been added yet.
* The default pipeline doesn't support gradients yet. The one that does is in test_grad, but it's a bit wonky (likely responsible for one of the xfails).
* The grad transform isn't very well documented.
* J/Jinv won't work with the debug VM right now.

abergeron commented on myia/prim/grad_implementations.py:
```diff
@@ -0,0 +1,245 @@
+    """Given a function for the bprop, make the augmented function."""
+    info = NamedDebugInfo(prim=prim, name=prim.name)
+
+    bprop = parse(fn)
```
Add a clone here, otherwise you'll get manager errors.  Also because you modify the returned graph.

breuleux replied:
The `clone(outer)` at the end of the function will clone `bprop` along with it, but it's true that it's better to clone a graph before modifying it, so I'll fix that.

lamblin replied:
OK.

abergeron commented on tests/test_grad.py:
```diff
@@ -0,0 +1,403 @@
+
+
+@pytest.mark.xfail(
+    reason="I think the opts to eliminate j/jinv are messing this one up."
```
If this is true then the opts should be fixed or disabled.

breuleux replied:
I disabled the opts and added support for J/Jinv in py_implementations.

abergeron commented on myia/debug/finite_diff.py:
```diff
@@ -0,0 +1,270 @@
+
+    def __init__(self, value):
+        """Initialize NoTestGrad."""
+        self.value = value
```
Support for NoTestGrad seems to make this ridiculously more complex than it should be.

breuleux commented:
I fixed one of the xfails and added support for J/Jinv in py_implementations. It seems to work.

breuleux commented:
That would be a good idea, yeah. There is some documentation in another branch which may still be helpful, but it needs to be updated for the new code: https://github.com/breuleux/myia/blob/gradient_transform/docs/gradient.ipynb

I'll try to add more comments and explanations.

breuleux commented:
I have added more comments and explanations, and I added a jupyter doc (GitHub renders it nicely):

https://github.com/breuleux/myia/blob/gradgen/docs/gradient.ipynb

lamblin commented on myia/grad.py:
```diff
@@ -0,0 +1,657 @@
+def J(graph: Graph, manager):
+    """Implement J on a Graph."""
+    if graph.transforms.get('grad', None):
+        return graph.transforms['grad']
```
Would that risk to cause perturbation confusion if `J` is called more than once?

breuleux replied:
No, this is just to ensure a graph is only transformed once. Different calls of the result of `J(f)` will compute their perturbations separately.

lamblin replied:
OK.

lamblin commented on tests/test_grad.py:
```diff
@@ -0,0 +1,415 @@
+
+@grad_test((4.0, 5.0), (6.4, -7.8))
+def test_if2(a, b):
+    # This is max, but what this is really testing is the most basic
```
This is not really `max` any more, I think :)

breuleux replied:
Indeed, it is not.

breuleux commented:
@lamblin I'm not sure what you mean by "add a test with a grad() in the original function".

lamblin commented:
I mean, a test decorated by `@grad_test`, like the other ones, but where the decorated function calls `grad`  (or something equivalent) already. That would be a test for a second-order derivative.
I understand if we are not there yet, though.

breuleux commented:
Oh, I see. It can't work with `@grad_test` since the Python version of the function is run to estimate derivatives, and it is not able to deal with `grad` right now. I tried to compile the graph for `grad(grad(f))` on some simple function, but it doesn't work. I roughly know what the problem is, but I'd rather work on that a bit later.

breuleux commented:
@lamblin @bartvm Is it OK to merge this?
